### ISSUE TITLE
Return value for client methods

### ISSUE THREAD

**Issue by flipchart** [SignalR/SignalR#1149]
Hub methods on the server are able to have return values for the client to consume, but not the other way around. This means that the server cannot call a client and get a value from the method invocation. It would be great if this could be added as it would make server and client more symmetrical in their abilities.

**Comment by fjmoreira** [comment]
I have the same question...

**Comment by timtos** [comment]
Good question. How would I mimic this in my app code? Any hints?

**Comment by ericgrover** [comment]
All you would need to do is have a call back method defined in your hub, so that the client instead of returning a value per se, would be calling a new method on the hub and passing the "return" value as the method parameter.  You could then bundle this all together with the TPL so that both the initial call and the call back could be awaited using the EAP pattern as illustrated here:  https://github.com/SignalR/SignalR/issues.

**Comment by victorperez2911** [comment]
OK

I tried the following way.

**On Server**
```C#
public class MyHub : Hub
    {
        private static readonly ConcurrentDictionary<Guid, object> ClientResponses = new ConcurrentDictionary<Guid, object>();

        public bool SendClientResponse(Guid Identificador, Object Response)
        {
            return ClientResponses.TryUpdate(Identificador, Response, null);
        }

        public TResponse GetClientResponse<TResponse>(Guid Identificador, int Attempts = 5, int SecondsBetweenAttempts = 1)
        {
            object _response = null;
            int _attempts = 0;
            TResponse _retorno;

            do
            {
                if (!ClientResponses.TryGetValue(Identificador, out _response))
                {
                    break;
                }

                _attempts += 1;

                if (_response == null && _attempts < Attempts)
                {
                    System.Threading.Thread.Sleep(SecondsBetweenAttempts*1000);
                }


            } while (_response == null && _attempts <= Attempts);

            bool remocao = ClientResponses.TryRemove(Identificador, out _response);

            _retorno = (TResponse)_response;

            return _retorno;

        }

        public string GetIpClientLoja()
        {
            Guid _identificador = Guid.NewGuid();
            ClientResponses.TryAdd(_identificador, null);
            Clients.Client(ConnectionIdClientLoja).GetIp(_identificador);
            return GetClientResponse<string>(_identificador);
        }
}
```


**On Client 1**
```C#
public static async Task GetIP(Guid Identificador)
        {
            //System.Threading.Thread.Sleep(100000);
            string _response = "192.168.13.200";
            bool retorno = await _proxy.Invoke<bool>("SendClientResponse", Identificador, _response);
            if (!retorno) Console.WriteLine("Erro ao Responder a Solicitação do Servidor Central");
        }
```

**On Client 2**
```C#
string ip = _proxy.Invoke<string>("GetIpClientLoja").Result;
Console.WriteLine("Resultado:");
Console.WriteLine(ip);
Console.WriteLine("Tecle qualquer coisa para continuar....");

```

First observation

It works.

Questions

1) Is there any better (and safer) way of doing this from the point of view of using threads?

2) An improvement would be welcome, I would like to optionally have the "SendClientResponse" method only terminate, and release the client if a response consumption method was correctly executed from the Server Side, is this possible?


Victor Perez

**Comment by davidfowl** [comment]
@victorperez2911 You should be using a TaskCompletionSource instead of sleeping waiting for results. What you have is extremely inefficient.

**Comment by victorperez2911** [comment]
@davidfowl 

Thanks For The Help

Made this change to the side of the Hub, what do you think?

```C#
public class MyHub : Hub
    {
        private static readonly ConcurrentDictionary<Guid, object> ClientResponses = new ConcurrentDictionary<Guid, object>();
        
        public bool SendClientResponse(Guid Identificador, Object Response)
        {
            return ClientResponses.TryUpdate(Identificador, Response, null);
        }

        public TResponse GetClientResponse<TResponse>(Guid Identificador, int Attempts = 5, int SecondsBetweenAttempts = 1)
        {
            object _response = null;
            int _attempts = 0;
            TResponse _retorno;
            do
            {
                if (!ClientResponses.TryGetValue(Identificador, out _response))
                {
                    break;
                }

                _attempts += 1;

                if (_response == null && _attempts < Attempts)
                {
                    System.Threading.Thread.Sleep(SecondsBetweenAttempts*1000);
                }

            } while (_response == null && _attempts <= Attempts);

            bool remocao = ClientResponses.TryRemove(Identificador, out _response);

            _retorno = (TResponse)_response;

            return _retorno;

        }

        public Task<string> GetIpClientLoja()
        {
            
            Guid _identificador = Guid.NewGuid();
            ClientResponses.TryAdd(_identificador, null);

            Clients.Client(ConnectionIdClientLoja).GetIp(_identificador);

            return getClientResponseAsync<string>(_identificador);
        }

        private Task<TResponse> getClientResponseAsync<TResponse>(Guid Identificador)
        {
            TaskCompletionSource<TResponse> tcs = new TaskCompletionSource<TResponse>();

            tcs.SetResult(GetClientResponse<TResponse>(Identificador));
            
            return tcs.Task;
        }
}
```

**Comment by davidfowl** [comment]
@victorperez2911 I still see a sleep.

**Comment by victorperez2911** [comment]
ok @davidfowl rrsrsrs

And now ?

```C#
public class MyHub : Hub
    {

        //Static  ?
        public static event EventHandler<GetClientResponseEventArgs> OnGetClientResponse;

        public void SendClientResponse(Guid Identificador, string Response)
        {
            //using ? make this Thread-Safe ?
            OnGetClientResponse?.Invoke(this, new GetClientResponseEventArgs(Identificador, Response));
        }

        public Task<string> GetIpClientLoja()
        {
            TaskCompletionSource<string> tcs = new TaskCompletionSource<string>();

            Guid _identificador = Guid.NewGuid();

            OnGetClientResponse += (s, args) => {
                if (args.Identificador != _identificador) return; // Do I need this for many clients?
                tcs.TrySetResult(args.Response);
            };

            Clients.Client(ConnectionIdClientLoja).GetIp(_identificador);

            return tcs.Task;
        }

        public class GetClientResponseEventArgs : EventArgs
        {
            public GetClientResponseEventArgs(Guid Identificador, string Response)
            {
                identificador = Identificador;
                response = Response;
            }

            private Guid identificador;
            private string response;

            public Guid Identificador
            {
                get { return identificador; }
            }

            public string Response
            {
                get { return response; }
            }
        }
}
```

**Comment by davidfowl** [comment]
Much better, but you can do better:

```C#
public class MyHub : Hub
{
    private static readonly ConcurrentDictionary<Guid, TaskCompletionSource<object>> ClientResponses = new ConcurrentDictionary<Guid, TaskCompletionSource<object>>();

    public void SendClientResponse(Guid identificador, string response)
    {
        TaskCompletionSource<object> tcs;
        if (ClientResponses.TryGetValue(out tcs))
        {
            // Trigger the task continuation
            tcs.TrySetResult(response);
        }
        else
        {
            // Client response for something that isn't being tracked, might be an error
        }
    }

    public async Task<string> GetIpClientLoja()
    {
        // Create an entry in the dictionary that will be used to track the client response
        var tcs = new TaskCompletionSource<string>();
        var identificador = Guid.NewGuid();

        ClientResponses.Add(identificador, tcs);

        // Call GetIp on the client passing the identifier
        Clients.Client(ConnectionIdClientLoja).GetIp(identificador);

        try 
        {
            // Wait for the client to respond
            // TODO: Cancel this task if the client disconnects (potentially by just adding a timeout)
            return await tcs.Task;

        }
        finally 
        {
            // Remove the tcs from the dictionary so that we don't leak memory
            ClientResponses.TryRemove(identificador, out tcs);
        }
    }
}
```

I left a TODO as an exercise for the reader 😄

**Comment by victorperez2911** [comment]
OMG!

I can not believe I was able to pull lines of code from you David :).

Following new version.

```C#
public class MyHub : Hub
    {

        
        private static readonly ConcurrentDictionary<Guid, TaskCompletionSource<object>> ClientResponses = new ConcurrentDictionary<Guid, TaskCompletionSource<object>>();

        public void SendClientResponse(Guid Identificador, string Response)
        {
            TaskCompletionSource<object> tcs;

            if (ClientResponses.TryGetValue(Identificador, out tcs))
            {
                // Trigger the task continuation
                tcs.TrySetResult(Response);
            }
            else
            {
                // Client response for something that isn't being tracked, might be an error
                //Test Only
                throw new Exception("Resposta não Esperada.");
            }
        }

        public async Task<string> GetIpClientLoja()
        {
            // Create an entry in the dictionary that will be used to track the client response
            var tcs = new TaskCompletionSource<object>();
            var identificador = Guid.NewGuid();

            ClientResponses.TryAdd(identificador, tcs);

            // Call GetIp on the client passing the identifier
            Clients.Client(ConnectionIdClientLoja).GetIp(identificador);

            try
            {
                // Wait for the client to respond
                // TODO: Cancel this task if the client disconnects (potentially by just adding a timeout)
                int timeout = 5000;
                var task = tcs.Task;
                if (await Task.WhenAny(task, Task.Delay(timeout)) == task)
                {
                    return (string)await task;
                }
                else
                {
                    return "TimedOut";//Only Test
                }
            }
            finally
            {
                // Remove the tcs from the dictionary so that we don't leak memory
                ClientResponses.TryRemove(identificador, out tcs);
            }
        }
}
```

**Comment by victorperez2911** [comment]
Hello @davidfowl 

Could you give me a hint?

I changed the method that calls the Client to be a bit more generic, and I don't have to redo the process of tracking down the answer with TaskCompletitionSource all the time, and it worked perfectly.
 ```C#
private async Task<TResponse> SendToClientAndWaitResponse<TResponse>(string ConnectionId, string Method, int SecondsWaitFromResponse, TResponse TimeOutResponse, params object[] parameters)
{
    Guid _identificador = Guid.NewGuid();

    TaskCompletionSource<object> tcs = new TaskCompletionSource<object>();

    bool responseTracked = ClientResponses.TryAdd(_identificador, tcs);

    IClientProxy proxy = Clients.Client(ConnectionId);

   //I combined that all methods that will be used with this must have mechanisms as last parameter the Guid for tracker.
    Array.Resize(ref parameters, parameters.Length + 1);
    parameters[parameters.Length - 1] = _identificador;

    await proxy.Invoke(Method, parameters);

    try
    {
        int timeout = SecondsWaitFromResponse * 1000;

        var task = tcs.Task;

        if (await Task.WhenAny(task, Task.Delay(timeout)) == task)
        {
            return (TResponse)await task;
        }
        else
        {
            return TimeOutResponse;
        }
    }
    finally
    {
        ClientResponses.TryRemove(_identificador, out tcs);
    }
}
```

But right now, I need to use this mechanism outside the Hub.

Is this possible?

An Extended Method of some type that can be accessed with the Context generated by the GetHubContext method would be great.

Thank you

**Comment by adityapadmakumar** [comment]
https://github.com/davidfowl/AspNetCoreDiagnosticScenarios/blob/master/AsyncGuidance.md?utm_content=buffer2137b&utm_medium=social&utm_source=twitter.com&utm_campaign=buffer#always-create-taskcompletionsourcet-with-taskcreationoptionsruncontinuationsasynchronously

**Comment by adityapadmakumar** [comment]
Always create TaskCompletionSource<T> with TaskCreationOptions.RunContinuationsAsynchronously
[https://github.com/davidfowl/AspNetCoreDiagnosticScenarios/blob/master/AsyncGuidance.md?utm_content=buffer2137b&utm_medium=social&utm_source=twitter.com&utm_campaign=buffer#always-create-taskcompletionsourcet-with-taskcreationoptionsruncontinuationsasynchronously](url)

**Comment by sixpeteunder** [comment]
Hi,

Amateur developer here. @victorperez2911 could you help me out with mine? I've tried several approaches and they do not seem to work. My use case is to get some additional data from the caller. Something is not working, and I'm not sure whether it is related to this or something else. What I currently have looks something like this:

## Server (Inspired by the code above)

```csharp
public class GameHub : Hub<IGameClient>
{
    private static readonly ConcurrentDictionary<Guid, TaskCompletionSource<Card>> _responses = new();

    public void RequestCard(Guid identifier, Card request)
    {
        // Trigger the task continuation if the tcs exists, else notify the user they waited too long
        if (_responses.TryGetValue(out  TaskCompletionSource<Card> tcs))   
            tcs.TrySetResult(response)
        else await Clients.Caller.ReceiveServerMessage("That request has timed out.");
    }

    public async Task PerformTurn()
    {
        var identifier = Guid.NewGuid();
        var tcs = new TaskCompletionSource<Card>();
        
        // Create an entry in the dictionary that will be used to track the client response
        _responses.TryAdd(identifier, tcs);

        // Call PromptForCardRequest on the caller passing the identifier
        await Clients.Caller.PromptForCardRequest(identifier);

        // If some condition passes, the caller can request a card.
        Card? request = null;
        try 
        {
            // Wait for the client to respond
            var task = await Task.WhenAny(tcs.Task, Task.Delay(TimeSpan.FromSeconds(10));
            if (task == tcs.Task) request = await task;
            else tcs.TrySetCancelled(); // Not too sure about this
        }
        finally 
        {
            // Remove the tcs from the dictionary so that we don't leak memory.
            _responses.TryRemove(identifier, out tcs);
        }

        if (request is not null) 
        {
            // Do something with request
        }
    }
}
```

## Client (Blazor Server application)

```csharp
hubConnection.On<Guid>("PromptForCardRequest", async identifier => {
    var request = PromptUserForCard();
    await hubConnection.SendAsync("RequestCard", identifier, request);
});
```

Most of this works, but for some reason, RequestCard is not being called. The most confusing part is that if I just have a button that calls `await hubConnection.SendAsync("RequestCard", Guid.NewGuid(), new Card())` , it works. What am I missing?

**Comment by davidfowl** [comment]
> In case anyone lands here from Google like I did, I fixed it by adding another hub (and connection) specifically for receiving (and sending) responses from the frontend. I'm not sure how efficient this is, but it was all I was able to find short of switching to streaming.

By default, we allow one concurrent call per connection to the hub. You can increase this to 2 to allow completions to work

```C#
services.AddSignalR().AddHubOptions<GameHub>(o =>
{
    o.MaximumParallelInvocationsPerClient = 2;
});
```

This will allow the client that is waiting for the invocation to also resume execution. 



The next thing is that you want to create the `TaskCompletionSource` with TaskCreationOptions.RunContinuationsAsynchronousl so that you don't create a dependency between the 2 hub methods. Read more about it here https://github.com/davidfowl/AspNetCoreDiagnosticScenarios/blob/master/AsyncGuidance.md#always-create-taskcompletionsourcet-with-taskcreationoptionsruncontinuationsasynchronously

**Comment by davidfowl** [comment]
This technique works fine when you only want to wait on the caller but gets much more complex when you want to wait on any arbitrary connection since you can't rely on the response coming back to the server that is waiting for it. That's where SignalR can do some magic.

**Comment by sixpeteunder** [comment]
Thank you @davidfowl! Where should I look if I want a deep dive on this sort of thing?

**Comment by davidfowl** [comment]
Experience with the failures 😄 is how I learn things like this. Past experience then helps with future designs. 

PS: There's another issue in the above logic. If the client disconnects then your `PerformTurn` will hang for 10 seconds. It would be cleaner to associate the incomplete turn with the connection so that if the client disconnects, you can also complete the pending operation. That also means responses doesn't need to be a static in the hub.

**Comment by sixpeteunder** [comment]
@davidfowl sorry to bother you again.

I have spotted the problem you've mentioned, where my code does not know that the client has disconnected, and hence continues to wait for the response, but I'm having some trouble figuring out exactly how to tie that in. Should I use the connectionId in the dictionary and then try to remove it in OnDisconnected?

**Comment by davidfowl** [comment]
@sixpeteunder yep. Instead of using a guid to track responses, you can use the connection id. This works if you can only have a single response outstanding at a time. Otherwise, you'll need to track connection ids and guids per connection id.


PS: [This comment](https://github.com/SignalR/SignalR/issues/1149#issuecomment-973887222) was about ASP.NET Core SignalR, not this version of SignalR my bad 😄.

**Comment by sixpeteunder** [comment]
> Most of this works, but for some reason, RequestCard is not being called. The most confusing part is that if I just have a button that calls await hubConnection.SendAsync("RequestCard", Guid.NewGuid().ToString(), new Card()) , it works. What am I missing?

I can't get this working. I'm back to my original issue (quoted below) when I do it like this:

```csharp
services.AddBlazorServer().AddHubOptions(o =>
{
    o.MaximumParallelInvocationsPerClient = 2;
});
```

I'm also creating `TaskCompletionSource` with `TaskCreationOptions.RunContinuationsAsynchronously` now. I did a minimal sample [here](https://github.com/sixpeteunder/parallel) to see whether it was any of my own code messing things up, but I got roughly the same result. `DoSomethingElse` is always called after 15seconds, no matter when I click on it. Is there anything specific to Blazor Server that I should do to get this working?

Also, I'm using ASP.NET 6 for both the original project and this sample. Didn't know this was for the Framework version 😅.
